cluster-classification/script.whizzml

```diff
@@ -0,0 +1,70 @@
+;; Given a cluster, creates a batch centroid and uses the output
+;; dataset where each instance is labeled by centroid to create an
+;; ensemble (random forest), and returns the importance of each field
+;; averaged over the ensemble.
+
+;; Given a cluster, returns a batchcentroid id created from that
+;; cluster whose output dataset is the original cluster dataset
+;; appended with the centroids the instances are part of.
+
+(define (label-by-cluster cl-id) 
+  (let (cl-ds (get-in (fetch cl-id) ["dataset"]) 
+        cl-fields (get-in (fetch cl-id) ["input_fields"]))
+    (create-and-wait-batchcentroid 
+     {"cluster" cl-id 
+      "dataset" cl-ds 
+      "output_dataset" true
+      "output_fields" cl-fields})))
+
+;; Given a list of pairs, creates the equivalent map
+(define (list-to-map nested-list)
+  (let (key-list (map head nested-list)
+        value-list (map (lambda (x) (nth x 1)) nested-list))
+    (make-map key-list value-list)))
+
+;; Given a function of two values and two maps, returns a new map
+;; where if a key appears in only one of the original maps its value
+;; is unchanged, but if it appears in both its value is the function
+;; of the two original values.  
+(define (merge-with fn map1 map2) 
+  (let (subset-keys (filter (lambda (x) (contains? map1 x)) (keys map2))
+        subset-values (map (lambda (x) (fn (get map1 x) (get map2 x))) 
+                           subset-keys) 
+        subset (make-map subset-keys subset-values)) 
+    (merge map1 (merge map2 subset))))
+
+;; Given a cluster-id, builds a random forest on centroid membership,
+;; and returns a paired list of field importance averaged across all
+;; models in the forest
+
+(define (cluster-classification cl-id)
+  (let (bc-id (label-by-cluster cl-id)
+        labeled (get-in (fetch bc-id) ["output_dataset_resource"])
+        _ (wait labeled)
+        cl-en (create-and-wait-ensemble {"dataset" labeled 
+                                         "randomize" true})
+        models (get-in (fetch cl-en) ["models"])
+        len-models (count models)
+        imp-lst (map (lambda (md-id) 
+                       (get-in (fetch md-id) ["model" "importance"])) 
+                     models)
+        imp-map (map list-to-map imp-lst)
+        imp-sum (reduce (lambda (x y) (merge-with + x y)) {} imp-map))
+    (delete labeled)
+    (delete bc-id)
+    (delete cl-en)
+    (map (lambda (x) (list (head x) (/ (nth x 1) len-models))) 
+         imp-sum)))
```

Comment:
and, regardless of what we finally do with map over maps, yours right
now is easier to write and read like this (this already works):

    (map (lambda (h t) [h (/ t len-models)]) (keys imp-sum) (values imp-sum))

since map can take multiple lists (and [x y] and (list x y) are equivalent).